### ISSUE TITLE
about "encoder training speed".

### ISSUE THREAD

**Issue by sunnnnnnnny** [CorentinJ/Real-Time-Voice-Cloning#59]
At present, the training speed is very slow. May I ask you to read all the data directly into the memory and then take it directly. How can I modify it?

**Comment by CorentinJ** [comment]
Yeah so the encoder training is definitely something that I need to work on. It's quite unusual in deep learning to have a loss function of cubic complexity w.r.t. the batch size, so it's not a problem you see often. What I should do is add a minor and a major batch size where the minor batch size is the number of speakers and the major one is the number of batches of such speakers. That would help leveraging all the available GPU memory.

As for the data loading, it can be a problem if your disk is slow. It's not possible to read all data in memory, the dataset is huge. And due to the nature of the training you can't prepare batches in advance either. You can check if you are throttled by disk reading speed by enabling the profiler in this line:
https://github.com/CorentinJ/Real-Time-Voice-Cloning/blob/master/encoder/train.py#L67

**Comment by sunnnnnnnny** [comment]
My computing memory is about 200G. I checked the training log. The data prepare time will take up more than 90% of the total time. Therefore, I consider reading the data into the memory first and reading it from the memory.
I also used the preload mode of nvidia's APEX framework to load, and the training speed is generally improved. I also used 3 card training, I found that the forward pass and the backward pass time have decreased, but there seems to be a problem, that is, the same batch_size=64 size is divided into 3 cards and the loss is calculated separately and the single card batch_size=64 is calculated. There is a difference in loss, do you think I am right?

**Comment by CorentinJ** [comment]
I haven't implemented any multi-gpu support so I have no clue.
